llvm/lib/SYCLLowerIR/PropagateAspectUsage.cpp

```diff
@@ -0,0 +1,492 @@
+                                     "shouldn't contain empty metadata nodes");
+
+    auto *TypeName = cast<MDString>(N.getOperand(0));
+    Type *T = StructType::getTypeByName(C, TypeName->getString());
+    assert(T &&
+           "invalid type referenced by intel_types_that_use_aspect metadata");
+
+    AspectsSetTy &Aspects = Result[T];
+    for (size_t I = 1; I != N.getNumOperands(); ++I) {
+      auto *CAM = cast<ConstantAsMetadata>(N.getOperand(I));
+      Constant *C = CAM->getValue();
+      Aspects.insert(cast<ConstantInt>(C)->getSExtValue());
+    }
+  }
+
+  return Result;
+}
+
+using TypesEdgesTy =
+    std::unordered_map<const Type *, std::vector<const Type *>>;
+
+/// Propagates aspects from type @Start to all types which
+/// are reachable by edges @Edges by BFS algorithm.
+/// Result is recorded in @Aspects.
+void propagateAspectsThroughTypes(const TypesEdgesTy &Edges, const Type *Start,
+                                  TypeToAspectsMapTy &Aspects) {
+  const AspectsSetTy &AspectsToPropagate = Aspects[Start];
+  SmallSetVector<const Type *, 16> TypesToPropagate;
+  TypesToPropagate.insert(Start);
+  for (size_t I = 0; I < TypesToPropagate.size(); ++I) {
+    const Type *T = TypesToPropagate[I];
+    Aspects[T].insert(AspectsToPropagate.begin(), AspectsToPropagate.end());
+    auto It = Edges.find(T);
+    if (It != Edges.end())
+      TypesToPropagate.insert(It->second.begin(), It->second.end());
+  }
+}
+
+/// Propagates given aspects to all types in module @M. Function accepts
+/// aspects in @TypesWithAspects reference and writes a result in this
+/// reference.
+/// Type T in the result contains an aspect A if Type T is a composite
+/// type (array, struct, vector) which contains elements/fields of
+/// another type TT, which in turn uses the aspect A.
+/// @TypesWithAspects argument consist of known types with aspects
+/// from metadata information.
+///
+/// The algorithm is the following:
+/// 1) Make a list of all structure types from module @M. The list also
+///    contains DoubleTy since it is optional as well.
+/// 2) Make from list a type graph which consists of nodes corresponding to
+///    types and directed edges between nodes. An edge from type A to type B
+///    corresponds to the fact that A is contained within B.
+///    Examples: B is a pointer to A, B is a struct containing field of type A.
+/// 3) For every known type with aspects propagate it's aspects over graph.
+///    Every propagation is a separate run of BFS algorithm.
+///
+/// Time complexity: O((V + E) * T) where T is the number of input types
+/// containing aspects.
+void propagateAspectsToOtherTypesInModule(
+    const Module &M, TypeToAspectsMapTy &TypesWithAspects) {
+  std::unordered_set<const Type *> TypesToProcess;
+  Type *DoubleTy = Type::getDoubleTy(M.getContext());
+
+  // 6 is taken from sycl/include/CL/sycl/aspects.hpp
+  static constexpr int AspectFP64 = 6;
```

Comment:
Do I understand it correctly that here we are injecting dependencies on external project to the compiler?
Imagine how hard it will be to debug bugs in this code if I update **only** `sycl/include/CL/sycl/aspects.hpp`.

Reply:
That is correct, both the compiler and the runtime must be in sync.

I don't really want to introduce a hard dependency between them and therefore I don't know what would be a good way to handle that. I suppose we should at least add a comment to `aspects.hpp` saying that if this file is changed, then `PorpagateAspectUsage` pass must also be changed, what do you think?

Reply:
We had an offline sync with @bader and agreed that this such dependency on runtime headers is a design issue. We decided not to block the PR and I've created #5892 to track the problem